cohttp/src/request.ml

```diff
@@ -87,44 +88,50 @@ let is_valid_uri path meth =
    | Some _ -> true
    | None -> not (String.length path > 0 && path.[0] <> '/'))
 
-let uri { resource ; headers ; meth ; _ } =
-  match resource with
-  | "*" ->
-    begin match Header.get headers "host" with
-    | None -> Uri.of_string ""
-    | Some host ->
-      let host_uri = Uri.of_string ("//"^host) in
-      let uri = Uri.(with_host (of_string "") (host host_uri)) in
-      Uri.(with_port uri (port host_uri))
-    end
-  | authority when meth = `CONNECT -> Uri.of_string ("//" ^ authority)
-  | path ->
-    let uri = Uri.of_string path in
-    begin match Uri.scheme uri with
-    | Some _ -> (* we have an absoluteURI *)
-      Uri.(match path uri with "" -> with_path uri "/" | _ -> uri)
-    | None ->
-      let empty = Uri.of_string "" in
-      let empty_base = Uri.of_string "///" in
-      let pqs = match Stringext.split ~max:2 path ~on:'?' with
-        | [] -> empty_base
-        | [path] ->
-          Uri.resolve "http" empty_base (Uri.with_path empty path)
-        | path::qs::_ ->
-          let path_base =
+let uri { scheme ; resource ; headers ; meth ; _ } =
+  let uri =
+    match resource with
+    | "*" ->
+      begin match Header.get headers "host" with
+      | None -> Uri.of_string ""
+      | Some host ->
+        let host_uri = Uri.of_string ("//"^host) in
+        let uri = Uri.(with_host (of_string "") (host host_uri)) in
+        Uri.(with_port uri (port host_uri))
```

Comment:
While I was working on this I noticed that this logic is could be simplified with `Uri.make` but I didn't want to over-complicate this PR so I left it alone.

This could be done like:

```suggestion
        let host_uri = Uri.of_string ("//"^host) in
        Uri.(make ?host:(host host_uri) ?port:(port host_uri) ()
```

Reply:
Feel free to add this change, looks good.